central/networkgraph/flow/datastore/internal/store/postgres/store.go

```diff
@@ -104,6 +106,36 @@ const (
 		(SELECT 1 from deployments parent WHERE child.Props_DstEntity_Id = parent.id::text AND parent.clusterid = $1)
 		AND Props_DstEntity_Type = 1
 		AND UpdatedAt < $2`
+
+	pruneNetworkFlowsReturnStmt = ` RETURNING child.Props_SrcEntity_Type, child.Props_SrcEntity_Id, child.Props_DstEntity_Type,
+		child.Props_DstEntity_Id, child.Props_DstPort, child.Props_L4Protocol,
+		child.LastSeenTimestamp, child.UpdatedAt, child.ClusterId::text;`
+
+	// The idea behind this statement is to prune orphan external (discovered)
+	// entities from the entities table. When flows are pruned using the above
+	// statements, the returned deleted flows are used to construct a list of
+	// deletion candidates for the external entities table, and then if any of
+	// those are no longer referenced by a network flow, they are deleted.
+	//
+	// As with flow pruning, this is performed in two queries (for src and dst entities)
+	// to improve query performance
+	pruneOrphanExternalNetworkEntitiesSrcStmt = `DELETE FROM network_entities entity
+	WHERE (entity.Info_Id = ANY($1)) AND
+	NOT EXISTS
+		(SELECT 1 FROM %s flow
+			WHERE flow.Props_SrcEntity_Type = 4
+			AND flow.Props_SrcEntity_Id = entity.Info_Id
+			AND entity.Info_ExternalSource_Discovered = true
+		);`
+
+	pruneOrphanExternalNetworkEntitiesDstStmt = `DELETE FROM network_entities entity
+	WHERE (entity.Info_Id = ANY($1)) AND
+	NOT EXISTS
+		(SELECT 1 FROM %s flow
+			WHERE flow.Props_DstEntity_Type = 4
+			AND flow.Props_DstEntity_Id = entity.Info_Id
+			AND entity.Info_ExternalSource_Discovered = true
+		);`
```

Comment:
Perhaps you can combine these two statements and pass in either "Props_SrcEntity_Id" or "Props_DstEntity_Id".

Reply:
I think it is easier to read and understand as it is.  IMO.

Reply:
I will have to merge the statements because there is a problem with the current implementation. Indeed, it reads as:

- delete all entities not used as a src
- then, delete all entities not used as a dst

We want to delete entities that are not used as dst nor src, and this can only be accomplished by combining the test in a single query.

Reply:
This has been combined here ->> https://github.com/stackrox/stackrox/pull/14769/commits/0573f22cb24a6eba1787bfa77590afe513facb49

Reply:
Were there any explain plans run on this?  The reason these are generally separate statements in network flows is performance.  The presence of that `OR` will very likely lead to full table scans.  I would like to see performance runs and/or explain plans of what existed before and after this commit.

Reply:
We may have to think through the logic if the `OR` is a problem and I suspect it is.

Reply:
Could a `UNION` work?

Reply:
The `OR` statement has been replaced with several negated conjunctions. As discussed, we will be spawning a long running cluster with a particular attention to the `rox_central_postgres_query_errors` metric and fix as necessary. In the worst scenario, the pruner can be disabled with the feature-flag.